go/vt/vtgate/safe_session.go

```diff
@@ -387,6 +387,13 @@ func (session *SafeSession) SetSystemVariable(name string, expr string) {
 	session.SystemVariables[name] = expr
 }
 
+// GetSystemVariables gets the system variables in the session.
+func (session *SafeSession) GetSystemVariables() map[string]string {
+	session.mu.Lock()
+	defer session.mu.Unlock()
+	return session.SystemVariables
```

Comment:
This is not safe. Note that there's a sibling method, `SetSystemVariable`, that can update the `session.SystemsVariables` map while holding `session.mu`. Here, you're only holding `session.mu` to read the value of `session.SystemVariables`, but you return the whole map to the caller, so the caller can then either _read a key from the map_ or _write a key to the map_ without holding the mutex. Neither of those things are safe to do because they can conflict with concurrent writes.

In order to have non-racy semantics, you'll need to create a hard copy of the map while you're holding `session.mu`, and then return that copy.

Reply:
We should probably just change this method to take the key in and return the value, and we can do the read while protected by the mutex - the mirror of the `SetSystemVariable` method :point_up:

Reply:
sounds wasteful to me, but maybe I'm overthinking it

another alternative would be a visitor pattern that executes under the protection of the mutex :shrug:

Reply:
Fixed via f0037678c9b7bf4156f7537f8be2a195554ca315 :)

Reply:
these settings are stored at the connection level, and MySQL protocol says that they should not be multiplexed. 
The purpose of the safeSession was actually in the execution phase where you are accessing same information for executing in parallel for multi-shards. 
The above information and SetSystemVariables are not in those paths. Ideally, we should just clean up these locks and only keep which are actually required.